lib/Dialect/ModArith/Conversions/ModArithToArith/ModArithToArith.cpp

```diff
@@ -14,23 +16,208 @@ namespace mod_arith {
 #define GEN_PASS_DEF_MODARITHTOARITH
 #include "lib/Dialect/ModArith/Conversions/ModArithToArith/ModArithToArith.h.inc"
 
-/// Returns a possibly extended modulus necessary to compute the given operation
-/// without overflow.
-template <typename ValueOrOpResult>
-TypedAttr modulusHelper(IntegerAttr mod, ValueOrOpResult op, bool mul = false,
-                        bool reduce = false) {
-  auto width = getElementTypeOrSelf(op).getIntOrFloatBitWidth();
-  auto modWidth = (mod.getValue() - 1).getActiveBits();
-  width = reduce ? width : std::max(width, mul ? 2 * modWidth : modWidth + 1);
+IntegerType convertModArithType(ModArithType type) {
+  APInt modulus = type.getModulus().getValue();
+  return IntegerType::get(type.getContext(), modulus.getBitWidth());
+}
+
+Type convertModArithLikeType(ShapedType type) {
+  if (auto modArithType = llvm::dyn_cast<ModArithType>(type.getElementType())) {
+    return type.cloneWith(type.getShape(), convertModArithType(modArithType));
+  } else {
+    return type;
+  }
+}
+
+class ModArithToArithTypeConverter : public TypeConverter {
+ public:
+  ModArithToArithTypeConverter(MLIRContext *ctx) {
+    addConversion([](Type type) { return type; });
+    addConversion(
+        [](ModArithType type) -> Type { return convertModArithType(type); });
+    addConversion(
+        [](ShapedType type) -> Type { return convertModArithLikeType(type); });
+
+    // Convert from an integer type to a mod arith type: use mreduce
+    addSourceMaterialization([](OpBuilder &builder, Type type,
+                                ValueRange inputs, Location loc) -> Value {
+      return builder.create<ReduceOp>(loc, type, inputs[0]).getResult();
+    });
+
+    // Convert from an integer type to a mod arith type: use mreduce
+    addArgumentMaterialization([](OpBuilder &builder, Type type,
+                                  ValueRange inputs, Location loc) -> Value {
+      return builder.create<ReduceOp>(loc, type, inputs[0]).getResult();
+    });
+
+    // Convert from a mod arith type to an integer type : use mextract
+    addTargetMaterialization([](OpBuilder &builder, Type type,
+                                ValueRange inputs, Location loc) -> Value {
+      return builder.create<ExtractOp>(loc, type, inputs[0]).getResult();
+    });
```

Comment:
I suspect these lines could be deleted without impacting the lowering at all; since all modarith ops are lowered by the patterns, any `unrealized_conversion_cast` ops that would be inserted by the absence of these lines would also be cleaned up.

Also, inserting `reduce` ops suggests nonzero work being applied (though cf. my question in the other comment about the semantics of reduce), and I see this type materialization stuff as temporary workarounds for the chicken-and-egg problem of dialect lowering. If, for some unexpected reason, these materialization ops stick around after the lowering is done, we'd prefer to be notified with an error (unresolved `unrealized_conversion_cast` error at the end of the lowering) than to have them quietly lowered to ops that use actual CPU cycles.

```suggestion
```